### PR TITLE
Add the --filter option for the restore command

### PR DESCRIPTION
**Pull request by andr-sokolov**
Add the --filter option for the restore command

Do not add a file into the processing queues, when the file should be filtered
out according to the JSON file from the --filter option. Check files from the
pg_default tablespace and non-built-in tablespaces. User relations can not be
placed in the pg_global tablespace, so this tablespace files are not checked.
Remove the unused attribute, because isRelationNeeded is used now.

### CODE REVIEW

File: src/command/restore/restore.c [hunk 1]
```diff
@@ -2073,6 +2074,32 @@ restoreProcessQueue(const Manifest *const manifest, List **const queueList)
             const ManifestFilePack *const filePack = manifestFilePackGet(manifest, fileIdx);
             const ManifestFile file = manifestFileUnpack(manifest, filePack);
 
+            Oid dbNode, relNode;
+            if (sscanf(strZ(file.name), "pg_data/base/%u/%u", &dbNode, &relNode) == 2)
+            {
+                if (!isRelationNeeded(dbNode, DEFAULTTABLESPACE_OID, relNode))
+                    continue;
+            }
+            else if (strBeginsWithZ(file.name, "pg_tblspc/"))
```
Comment (RekGRpth) [inline src/command/restore/restore.c]: Directly such strings are used only in tests. Try using ready-made macros like https://github.com/arenadata/pgbackrest/blob/3a86d4ea84fec499148cebf9144344d34de24fd8/src/command/restore/restore.c#L1285-L1286

File: src/command/restore/restore.c [hunk 2]
```diff
@@ -2073,6 +2074,32 @@ restoreProcessQueue(const Manifest *const manifest, List **const queueList)
...
+                StringList *lst = strLstNewSplitZ(file.name, "/");
+                if (strLstSize(lst) == 5)
+                {
+                    Oid spcNode;
+                    if (
+                        // Check all sscanf-s at once to simplify the tests
+                        sscanf(strZ(strLstGet(lst, 1)), "%u", &spcNode) +
+                        sscanf(strZ(strLstGet(lst, 3)), "%u", &dbNode) +
+                        sscanf(strZ(strLstGet(lst, 4)), "%u", &relNode) == 3 &&
+                        !isRelationNeeded(dbNode, spcNode, relNode))
```
Comment (RekGRpth) [inline src/command/restore/restore.c]: The `isRelationNeeded` function may throw an error and will not reach the point of freeing the list.

Comment (andr-sokolov) [reply src/command/restore/restore.c]: It is not a problem, because the program will be terminated in this case

Comment (RekGRpth) [reply src/command/restore/restore.c]: pgbackrest does not typically use explicit memory deallocation, instead it has contexts.

File: src/command/restore/restore.c [hunk 3]
```diff
@@ -2073,6 +2074,32 @@ restoreProcessQueue(const Manifest *const manifest, List **const queueList)
...
+                    {
+                        strLstFree(lst);
+                        continue;
+                    }
+                }
+                strLstFree(lst);
```
Comment (RekGRpth) [inline src/command/restore/restore.c]: may be add some comments like https://github.com/arenadata/pgbackrest/blob/3a86d4ea84fec499148cebf9144344d34de24fd8/src/info/manifest.c#L3214-L3224

File: src/command/restore/restore.c [hunk 4]
```diff
@@ -2073,6 +2074,32 @@ restoreProcessQueue(const Manifest *const manifest, List **const queueList)
...
+            }
+            else if (strBeginsWithZ(file.name, MANIFEST_TARGET_PGTBLSPC "/"))
+            {
+                StringList *lst = strLstNewSplitZ(file.name, "/");
+                if (strLstSize(lst) == 5)
+                {
+                    Oid spcNode;
+                    if (
+                        // Check all sscanf-s at once to simplify the tests
+                        sscanf(strZ(strLstGet(lst, 1)), "%u", &spcNode) +
```
Comment (RekGRpth) [inline src/command/restore/restore.c]: Can this line return 0? If so, why do we need to check the other lines?

File: src/command/restore/restore.c (hunk 3 above)
Comment (andr-sokolov) [reply src/command/restore/restore.c]: Done

File: src/command/restore/restore.c (hunk 4 above)
Comment (andr-sokolov) [reply src/command/restore/restore.c]: Yes, it can. The current solution is used to simplify the tests.

**Comment by KnightMurloc** [comment]
Do we want to apply the filter only to gpdb6 for now? Otherwise, we will be able to filter the files and fall when trying to filter the WAL due to the unsupported version.

File: src/command/restore/restore.c [hunk 5]
```diff
@@ -2073,6 +2074,34 @@ restoreProcessQueue(const Manifest *const manifest, List **const queueList)
             const ManifestFilePack *const filePack = manifestFilePackGet(manifest, fileIdx);
             const ManifestFile file = manifestFileUnpack(manifest, filePack);
 
+            Oid dbNode, relNode;
+            // If this file is located in the default tablespace
+            if (sscanf(strZ(file.name), MANIFEST_TARGET_PGDATA "/" PG_PATH_BASE "/%u/%u", &dbNode, &relNode) == 2)
```
Comment (RekGRpth) [inline src/command/restore/restore.c]: How much does the sscanf call slow down the execution? And why do we need it if the filter is not set at all?

File: src/command/restore/restore.c (hunk 4 above)
Comment (RekGRpth) [reply src/command/restore/restore.c]: But are tests the first priority?

File: test/src/module/command/restoreTest.c [hunk 6]
```diff
@@ -3392,6 +3392,109 @@ testRun(void)
 
         // Check that file was restored to full size with a partial write
         TEST_RESULT_LOG_EMPTY_OR_CONTAINS(", bi 128KB/256KB, ");
+        harnessLogLevelSet(logLevelWarn);
```
Comment (KnightMurloc) [inline test/src/module/command/restoreTest.c:3395]: I think it's worth moving this after `TEST_TITLE` so that it's clear which test it relates to.

File: test/src/module/command/restoreTest.c [hunk 7]
```diff
@@ -3392,6 +3392,109 @@ testRun(void)
...
+        HRN_STORAGE_PUT_EMPTY(storagePgWrite(), PG_PATH_BASE "/1/40045");
+        HRN_STORAGE_PUT_EMPTY(storagePgWrite(), PG_PATH_BASE "/1/40045_fsm");
+        HRN_STORAGE_PUT_EMPTY(storagePgWrite(), PG_PATH_BASE "/1/40045_vm");
+        HRN_STORAGE_PUT_EMPTY(storageTest, "ts/PG_15_202209061/16416/20000");
+        HRN_STORAGE_PUT_EMPTY(storageTest, "ts/PG_15_202209061/16416/20000_fsm");
+        HRN_STORAGE_PUT_EMPTY(storageTest, "ts/PG_15_202209061/16416/20000_vm");
+        HRN_STORAGE_PUT_EMPTY(storageTest, "ts/PG_15_202209061/16416/20001");
+        HRN_STORAGE_PUT_EMPTY(storageTest, "ts/PG_15_202209061/16416/20001_fsm");
+        HRN_STORAGE_PUT_EMPTY(storageTest, "ts/PG_15_202209061/16416/20001_vm");
+        HRN_STORAGE_PUT_EMPTY(storageTest, "ts/PG_15_202209061/16416/" PG_FILE_PGVERSION);
```
Comment (KnightMurloc) [inline test/src/module/command/restoreTest.c]: Add the `pg_filenode.map` file. To be sure that it is also restored. Because it is mentioned in the SRS.

File: src/command/restore/restore.c [hunk 8]
```diff
@@ -2073,6 +2074,34 @@ restoreProcessQueue(const Manifest *const manifest, List **const queueList)
...
+            {
+                StringList *lst = strLstNewSplitZ(file.name, "/");
+                if (strLstSize(lst) == 5)
+                {
+                    Oid spcNode;
+                    if (
+                        // Check all sscanf-s at once to simplify the tests
+                        sscanf(strZ(strLstGet(lst, 1)), "%u", &spcNode) +
+                        sscanf(strZ(strLstGet(lst, 3)), "%u", &dbNode) +
+                        sscanf(strZ(strLstGet(lst, 4)), "%u", &relNode) == 3 &&
```
Comment (RekGRpth) [inline src/command/restore/restore.c]: I don't quite understand, where is the test for these lines? I expected it to be something like `123/qwe/456/789`

File: test/src/module/command/restoreTest.c [hunk 9]
```diff
@@ -3456,45 +3471,67 @@ testRun(void)
...
-            "base/1/3\n"
             "base/1/40045\n"
             "base/1/40045_fsm\n"
             "base/1/40045_vm\n"
-            "base/1/44\n"
             "global/\n"
             "global/pg_control\n"
             PG_PATH_PGTBLSPC "/\n"
-            PG_PATH_PGTBLSPC "/16415>\n"
-            "postgresql.auto.conf\n",
```
Comment (KnightMurloc) [inline test/src/module/command/restoreTest.c]: Why were these files removed from the test?

File: src/command/restore/restore.c [hunk 10]
```diff
@@ -2473,6 +2502,12 @@ cmdRestore(void)
             strNewFmt(STORAGE_REPO_BACKUP "/%s/" BACKUP_MANIFEST_FILE, strZ(backupData.backupSet)), backupData.repoCipherType,
             backupData.backupCipherPass);
 
+        if (cfgOptionTest(cfgOptFilter) &&
+            (cfgOptionStrId(cfgOptFork) != CFGOPTVAL_FORK_GPDB || manifestData(jobData.manifest)->pgVersion != PG_VERSION_94))
+        {
+            THROW(OptionInvalidError, "option '" CFGOPT_FILTER "' is supported on GPDB 6 only");
+        }
```
Comment (RekGRpth) [inline src/command/restore/restore.c:2497]: may be something like
```suggestion
        if (cfgOptionTest(cfgOptFilter) &&
            (cfgOptionStrId(cfgOptFork) != CFGOPTVAL_FORK_GPDB || manifestData(jobData.manifest)->pgVersion != PG_VERSION_94))
        {
            if (cfgOptionStrId(cfgOptFork) != CFGOPTVAL_FORK_GPDB)
                THROW(OptionInvalidError, "option '" CFGOPT_FILTER "' is supported on GPDB only");
            if (manifestData(jobData.manifest)->pgVersion != PG_VERSION_94))
                THROW(OptionInvalidError, "option '" CFGOPT_FILTER "' is supported on GPDB 6 only");
        }
```

File: test/src/module/command/restoreTest.c [hunk 11]
```diff
@@ -3456,45 +3471,67 @@ testRun(void)
...
+        TEST_TITLE("restore filter is supported on GPDB 6 only");
+
+        argList = strLstNew();
+        hrnCfgArgRawZ(argList, cfgOptStanza, "test1");
+        hrnCfgArgRaw(argList, cfgOptRepoPath, repoPath);
+        hrnCfgArgRaw(argList, cfgOptPgPath, pgPath);
+        hrnCfgArgRawZ(argList, cfgOptSpoolPath, TEST_PATH "/spool");
+        hrnCfgArgRawZ(argList, cfgOptFilter, TEST_PATH "/restore_filter.json");
+        hrnCfgArgRawZ(argList, cfgOptRepoCipherType, "aes-256-cbc");
+        hrnCfgEnvRawZ(cfgOptRepoCipherPass, TEST_CIPHER_PASS);
```
Comment (KnightMurloc) [inline test/src/module/command/restoreTest.c:3524]: It looks like these options have been removed from previous tests. Why were they left here?

Comment (andr-sokolov) [reply test/src/module/command/restoreTest.c:3524]: > It looks like these options have been removed from previous tests. Why were they left here?

The test1 stanza is created with cipher, so these two options are necessary to restore from it

File: test/src/module/command/restoreTest.c (hunk 6 above)
Comment (andr-sokolov) [reply test/src/module/command/restoreTest.c:3395]: logLevelDetail is set in the beginning of the last existing test. We should set default log level at the end of the test

File: src/command/restore/restore.c (hunk 10 above)
Comment (andr-sokolov) [reply src/command/restore/restore.c:2497]: I think that `GPDB 6 only` is suitable for the both cases and more informative. The code will be simpler: one THROW instead of two

Comment (RekGRpth) [reply src/command/restore/restore.c:2497]: But to add support for the 7X, you will have to change less.

File: src/command/restore/restore.c [hunk 12]
```diff
@@ -2068,11 +2069,40 @@ restoreProcessQueue(const Manifest *const manifest, List **const queueList)
...
+                }
+                // If this file is located in a non-built-in tablespace
+                else if (sscanf(strZ(file.name), MANIFEST_TARGET_PGTBLSPC "/%u", &spcNode) == 1)
+                {
+                    StringList *lst = strLstNewSplitZ(file.name, "/");
+                    if (strLstSize(lst) == 5)
+                    {
+                        if (
+                            sscanf(strZ(strLstGet(lst, 3)), "%u", &dbNode) == 1 &&
+                            sscanf(strZ(strLstGet(lst, 4)), "%u", &relNode) == 1 &&
```
Comment (RekGRpth) [inline src/command/restore/restore.c]: I don't quite understand, where is the test for these lines? I expected it to be something like 123/qwe/456/789

Comment (andr-sokolov) [reply src/command/restore/restore.c]: https://github.com/arenadata/pgbackrest/pull/44/files#diff-6d9ab887a8ce7469b6f93279461cc94e5a2f4d43bbaaba5287b5caa94ac8ff4bR3414-R3415

File: test/src/module/command/restoreTest.c (hunk 7 above)
Comment (andr-sokolov) [reply test/src/module/command/restoreTest.c]: done

File: src/command/restore/restore.c [hunk 13]
```diff
@@ -2068,11 +2069,28 @@ restoreProcessQueue(const Manifest *const manifest, List **const queueList)
...
+            if (isFilterSet)
+            {
+                Oid dbNode, relNode;
+                Oid spcNode = DEFAULTTABLESPACE_OID;
+                char tsId[32];
+                if (
+                    // If this file is located in the default tablespace
+                    sscanf(strZ(file.name), MANIFEST_TARGET_PGDATA "/" PG_PATH_BASE "/%u/%u", &dbNode, &relNode) == 2 ||
+                    // If this file is located in a non-built-in tablespace
+                    sscanf(strZ(file.name), MANIFEST_TARGET_PGTBLSPC "/%u/%31[^/]/%u/%u", &spcNode, tsId, &dbNode, &relNode) == 4)
```
Comment (KnightMurloc) [inline src/command/restore/restore.c]: an unused value may be skipped.
```suggestion
                    sscanf(strZ(file.name), MANIFEST_TARGET_PGTBLSPC "/%u/%*31[^/]/%u/%u", &spcNode, &dbNode, &relNode) == 4)
```

Comment (andr-sokolov) [reply src/command/restore/restore.c]: done

**Review by KnightMurloc — APPROVED** [review]

**Review by RekGRpth — APPROVED** [review]
